mantidproject/mantid
# Stop SaveHKL recalculating tbar if it's already in PeaksWorkspace

**Description of work.**

I've changed the SaveHKL algorithm so it only calculates tbar for general sample shapes if the corresponding column in the peaks workspace is zero (ie empty) for all peaks. The column in the peaks workspace can be populated by running AddAbsorptionWeightedPathLengths.

As part of this I have also:
- moved the "single path" calculation of tbar for general shapes out of SaveHKL and into the AddAbsorptionWeightedPathLengths algorithm
- made AddAbsorptionWeightedPathLengths use the same parallel for loop that is in MonteCarloAbsorption. This will make it quicker which is generally useful - but also helpful if calling this algorithm from SaveHKL
- moved the AppendFile logic to the end of SaveHKL::exec so that the OutputWorkspace includes any changes made to the peaks during SaveHKL (such as the tbar calculation, but also the anvred corrections)
- moved MCInteractionVolume::calculateAbsorption onto the Track class so it's accessible to people not doing a Monte Carlo simulation (and renamed it to calculateAttenuation because it includes the scattering cross section as well as absorption)

**To test:**

Run this script and check that the data in the tbar column (column 8) is different in the files MyPeaks.hkl and MyPeaksPreCalculatedTbar.hkl files. The files differ because the call to AddAbsorptionWeightedPathLengths in the script below runs a Monte Carlo calculation of tbar instead of the simple single path calculation that is done by SaveHKL. The transmission in column 11 (which is closely related to tbar) will also be different:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

import os

#load a peaks workspace from file
peaks = LoadIsawPeaks(Filename=r'Peaks5637.integrate')
print("Number of peaks in table {}".format(peaks.rowCount()))

path = os.path.join(os.path.expanduser("~"), "MyPeaks.hkl")
SetSample(peaks,
          Geometry={'Shape': 'Cylinder', 'Height': 4.0,
                    'Radius': 0.8,
                    'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'V', 'SampleNumberDensity': 0.1})
SaveHKL(peaks, path, MinWavelength=0.5, MaxWavelength=2,MinDSpacing=0.2, SortBy='Bank')
AddAbsorptionWeightedPathLengths(InputWorkspace=peaks)
path = os.path.join(os.path.expanduser("~"), "MyPeaksPreCalculatedTbar.hkl")
SaveHKL(peaks, path, MinWavelength=0.5, MaxWavelength=2,MinDSpacing=0.2, SortBy='Bank')
```

Fixes #29954.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
